lib/invoker/dns_cache.rb

```diff
@@ -2,22 +2,84 @@ module Invoker
   class DNSCache
     attr_accessor :dns_data
 
+    # INI:
+    # [app]
+    # port = 8000
+    # location = www2.app/blah
+    #
+    # [chat]
+    # port = 8001
+    # location www.app/chat
+    #
+    # [api]
+    # port = 8002
+    # location = www.app/api
+    #
+    # dns_data:
+    # {
+    #   'api' => [
+    #     ['', 8002]
+    #   ],
+    #   'app' => [
+    #     ['', 8000]
+    #   ],
+    #   'chat' => [
+    #     ['', 8001]
+    #   ],
+    #   'www.app' => [
+    #     ['', 8000],
+    #     ['/api, 8002],
+    #     ['/chat, 8001]
+    #   ],
+    #   'www2.app' => [
+    #     ['/blah', 8000]
+    #   ]
+    # }
     def initialize(config)
-      self.dns_data = {}
+      self.dns_data = Hash.new {|h,k| h[k] = []}
       @dns_mutex = Mutex.new
       Invoker.config.processes.each do |process|
         if process.port
-          dns_data[process.label] = { 'port' => process.port }
+          add(process.label, process.port)
+          if process.location
+            process.location.split(' ').each do |loc|
+              add(loc, process.port)
+            end
+          end
         end
       end
     end
 
-    def [](process_name)
-      @dns_mutex.synchronize { dns_data[process_name] }
+    def find_process(host, path)
+      @dns_mutex.synchronize {
+        until dns_data.include?(host) || host.nil?
+          host = host.split('.', 2)[1]
+        end
+
+        dns_data[host].reverse_each {|prefix, port|
+          if path_matches_prefix?(path, prefix)
+            return {:port => port}
+          end
+        }
+
+        nil
+      }
+    end
+
+    def add(location, port)
+      @dns_mutex.synchronize {
+        host, path_prefix = split_host_path(location)
+        (dns_data[host] << [path_prefix.to_s, port]).sort_by! &:length
+      }
+    end
+
+    private
+    def split_host_path(label)
+      label.split(%r{(?=/)}, 2)
```

Comment:
Ah, so this seems to be supporting both old and new format. Do you think it is best to support one or another? I would rather just have separate `location` field in config as proposed. Supporting both `location` and `path` in process label is somewhat unnecessary imo.